Schedule task on buffered PerformanceObserver

francovs opened the pull request:
#### fbf81e4f230b7c80faa2fe9d8103418154bac661
<pre>
Schedule task on buffered PerformanceObserver
<a href="https://bugs.webkit.org/show_bug.cgi?id=297765">https://bugs.webkit.org/show_bug.cgi?id=297765</a>
<a href="https://rdar.apple.com/158950657">rdar://158950657</a>

Reviewed by Chris Dumez and Ryan Reno.

PerformanceObserver::observe() with `buffered: true` now schedules a task to
run all callbacks instead of invoking the relevant callbacks directly. The
new behavior is more conformant; see
<a href="https://www.w3.org/TR/performance-timeline/#observe-method">https://www.w3.org/TR/performance-timeline/#observe-method</a>, step 7.5.3.

This change exposed some inconsistencies in the logic that sends navigation
entries to PerformanceObservers, so a minor rework was required. In the new
implementation, the navigation entry is only exposed to PerformanceObservers
after the load event is handled. This is done in
LocalDOMWindow::dispatchLoadEvent(). The previous implementation did this
in FrameLoader::checkLoadCompleteForThisFrame(), with less consistent behavior.
Notably, web-platform-tests/navigation-timing/po-navigation.html seems to
fail randomly if opened outside of testing.

The spec is unclear with regards to when a navigation performance entry
should be considered available. In this implementation, there is some span
of time before the onload event in which the navigation entry is available
through performance.getEntries() while not being observable through
PerformanceObservers. This inconsistency is reasonable: the navigation entry
contains timing information about the onload event itself, so can only be
considered complete after that. Considering PerformanceObservers were
introduced to reduce unnecessary processing, it makes sense for them to
only receive completed entries.

New tests added:
* LayoutTests/imported/w3c/web-platform-tests/performance-timeline/buffered-does-not-sync-invoke.html :
    verifies that calling observe() does not invoke the callbacks synchronously. This is a direct test
    of the bug that motivated the patch.
* LayoutTests/performance-api/performance-observer-no-duplicate-navigation.html : exercises a latent
    bug (now removed) in which navigation entries were queued twice

Canonical link: <a href="https://commits.webkit.org/299508@main">https://commits.webkit.org/299508@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8346ac6842f68ee3fbdf9636f4190c7c75db8f21

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29174 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125031 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70904 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120720 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47105 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90193 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59706 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31256 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106547 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70700 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30318 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24661 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68690 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100699 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24848 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128083 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45749 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34548 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98848 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46115 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102767 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98628 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25146 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44075 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22078 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42306 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45619 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51297 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45084 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48429 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46769 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on LayoutTests/performance-api/performance-observer-buffered-schedules-task.html:
```diff
@@ -0,0 +1,31 @@
+<!DOCTYPE HTML>
+<html>
+<head>
+<script src="../resources/js-test-pre.js"></script>
```
Just use `js-test.js` and then you can get rid of `pre and post` scripts.

rreno reviewed (commented):
In the commit message
> with less consistend behavior.

should be "consistent"

rreno commented on Source/WebCore/page/Performance.h:
```diff
@@ -109,6 +109,7 @@ class Performance final : public RefCounted<Performance>, public ContextDestruct
 
     void addNavigationTiming(DocumentLoader&, Document&, CachedResource&, const DocumentLoadTiming&, const NetworkLoadMetrics&);
     void navigationFinished(const NetworkLoadMetrics&);
+    void navigationFullyFinished();
```
A deeper refactor might be a good follow-up but as this patch is `navigationFinished` and `navigationFullyFinished` are not great names and it's not clear what the difference is in terms of usage. I think making the names more distinct and clear is a good thing to do here.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3bc8db6028923ec074a3a32aeaa4f4ba975b2159. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49790

rreno commented on Source/WebCore/page/PerformanceNavigationTiming.cpp:
```diff
@@ -140,7 +140,7 @@ double PerformanceNavigationTiming::duration() const
 }
 
-void PerformanceNavigationTiming::navigationFinished(const NetworkLoadMetrics& metrics)
+void PerformanceNavigationTiming::updateNetworkLoadMetrics(const NetworkLoadMetrics& metrics)
```
It's probably better to call this `documentLoadFinished` to match the caller and make it clear what the update implies - that something has finished.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8346ac6842f68ee3fbdf9636f4190c7c75db8f21. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49790

francovs committed:
Schedule task on buffered PerformanceObserver
https://bugs.webkit.org/show_bug.cgi?id=297765
rdar://158950657

Reviewed by Chris Dumez and Ryan Reno.

PerformanceObserver::observe() with `buffered: true` now schedules a task to
run all callbacks instead of invoking the relevant callbacks directly. The
new behavior is more conformant; see
https://www.w3.org/TR/performance-timeline/#observe-method, step 7.5.3.

This change exposed some inconsistencies in the logic that sends navigation
entries to PerformanceObservers, so a minor rework was required. In the new
implementation, the navigation entry is only exposed to PerformanceObservers
after the load event is handled. This is done in
LocalDOMWindow::dispatchLoadEvent(). The previous implementation did this
in FrameLoader::checkLoadCompleteForThisFrame(), with less consistent behavior.
Notably, web-platform-tests/navigation-timing/po-navigation.html seems to
fail randomly if opened outside of testing.

The spec is unclear with regards to when a navigation performance entry
should be considered available. In this implementation, there is some span
of time before the onload event in which the navigation entry is available
through performance.getEntries() while not being observable through
PerformanceObservers. This inconsistency is reasonable: the navigation entry
contains timing information about the onload event itself, so can only be
considered complete after that. Considering PerformanceObservers were
introduced to reduce unnecessary processing, it makes sense for them to
only receive completed entries.

New tests added:
* LayoutTests/imported/w3c/web-platform-tests/performance-timeline/buffered-does-not-sync-invoke.html :
    verifies that calling observe() does not invoke the callbacks synchronously. This is a direct test
    of the bug that motivated the patch.
* LayoutTests/performance-api/performance-observer-no-duplicate-navigation.html : exercises a latent
    bug (now removed) in which navigation entries were queued twice

Canonical link: https://commits.webkit.org/299508@main
